lib/src/clojure_lsp/dep_graph.clj

```diff
@@ -0,0 +1,344 @@
+                                        clojure.edn              1
+                                        clojure.java.io          1
+                                        clojure.string           1
+                                        clojure.tools.cli        1
+                                        pod.clojure-lsp.api      1}
+                       :files          #{"~/code/clojure-lsp/cli/src/clojure_lsp/main.clj"}
+                       :internal?      true}
+    clojure.tools.cli {:dependents     {clj-depend.main                                    1
+                                        cljfmt.main                                        1
+                                        clojure-lsp.main                                   1
+                                        clojure.tools.deps.alpha.script.generate-manifest2 1
+                                        clojure.tools.deps.alpha.script.make-classpath2    1
+                                        clojure.tools.deps.alpha.script.print-tree         1
+                                        clojure.tools.deps.alpha.script.resolve-tags       1
+                                        kaocha.runner                                      1}
+                       :aliases        {cli 7, nil 1}
+                       :from-internal? true
+                       :from-langs     {:clj 8}
+                       :dependencies   {clojure.string 2, goog.string.format 1}
+                       :files          #{"~/.m2/repository/org/clojure/tools.cli/1.0.206/tools.cli-1.0.206.jar:clojure/tools/cli.cljc"}
+                       :internal?      false}}
+  (-> @db/db*
+      :file-meta
+      (select-keys ["~/code/clojure-lsp/cli/src/clojure_lsp/main.clj"
+                    "~/.m2/repository/org/clojure/tools.cli/1.0.206/tools.cli-1.0.206.jar:clojure/tools/cli.cljc"]))
+
+  '{"~/code/clojure-lsp/cli/src/clojure_lsp/main.clj"
+    {:internal? true
+     :langs #{:clj}
+     :namespaces #{clojure-lsp.main}}
+    "~/.m2/repository/org/clojure/tools.cli/1.0.206/tools.cli-1.0.206.jar:clojure/tools/cli.cljc"
+    {:internal? false
+     :langs #{:clj :cljs}
+     :namespaces #{clojure.tools.cli}}})
+
+;;;; DB Maintenance
+
+(def ^:private empty-multiset {})
+
+(defn ^:private mulitset-conj [ms x]
+  (update ms x (fnil inc 0)))
+
+(defn ^:private multiset-disj [ms x]
+  (if-let [v (get ms x)]
+    (let [new-v (dec v)]
+      (if (zero? new-v)
+        (dissoc ms x)
+        (assoc ms x new-v)))
+    ms))
+
+(defn ms-distinct [ms] (some-> ms keys))
+
+(defn ms-overlaps-set? [ms s]
+  ;; Conceptually set overlap is symmetric, but with our implementation of
+  ;; multiset, in practice we must check whether the items of the set are in the
+  ;; multiset, not vice-versa.
+  (boolean (and ms (some ms s))))
+
+(def ^:private ms-conj (fnil mulitset-conj empty-multiset))
+(def ^:private ms-disj (fnil multiset-disj empty-multiset))
```

Comment:
Doing this now would be overengineering but I think it's worth to mention so we can be aware of this if need more performance: Since this is critical performance code for clojure-lsp, we could consider having a `Multiset` defrecord so we could have more performance probably to work on that data structure and a place for those functions

Reply:
There are a few libraries that implement multisets as defrecords. But those libraries do a lot of other stuff—multiset versions of union, intersection, etc.—which we don't need. I didn't want to add a dependency if we weren't going to use much of its code.

FWIW, those libraries use defrecords so that they can implement various protocols. For example, by implementing `clojure.lang.IPersistentCollection`, you can call `(clojure.core/conj ms item)`, instead of what I have here `(mulitiset-conj ms item)`. But under the hood the defrecords in those libraries wrap a hashmap, which maps keys to frequencies, just like I've done here. So that extra layer of indirection might actually be slower. Anyway, if our goal was performance, we'd have test defrecords carefully.

Reply:
Sure, if we need in the future, we could start with out own defrecord simple implementation just encapsulating those functions, but something not important for now